rig/machine_control/regions.py

```diff
@@ -144,63 +117,82 @@ def __init__(self, base_x=0, base_y=0, level=0):
-            for e in self.locally_selected:
-                elements |= 1 << e
-            yield (region_code | elements)
+        # Generate core masks for any regions which are selected at this level
+        # Create a mapping from subregion mask to core numbers
+        subregions_cores = collections.defaultdict(lambda: 0x0)
+        for core, subregions in enumerate(self.locally_selected):
+            if subregions:  # If any subregions are selected on this level
+                subregions_cores[subregions] |= 1 << core
 
-        # Include subregions if they exist
-        if self.level < 3:
-            for i, sr in enumerate(self.subregions):
-                if i not in self.locally_selected and sr is not None:
-                    for r in sr.get_regions():
-                        yield r
+        # Order the locally selected items and then yield them
+        sorted_subregion_coremasks = sorted(list(subregions_cores.items()),
+                                            key=lambda x: (x[0] << 32) | x[1])
+        for (subregions, coremask) in sorted_subregion_coremasks:
+            yield (region_code | subregions), coremask
 
-    def add_coordinate(self, x, y):
-        """Add a new coordinate to the region tree.
+        if self.level < 3:
+            # Iterate through the subregions and recurse, we iterate through in
+            # the order which ensures that anything we yield is in increasing
+            # order.
+            for i in (4*x + y for y in range(4) for x in range(4)):
+                subregion = self.subregions[i]
+                if subregion is not None:
+                    for x in subregion.get_regions_and_coremasks():
+                        yield x
+
+    def add_core(self, x, y, p):
+        """Add a new core to the region tree.
 
         Raises
         ------
         ValueError
-            If the co-ordinate is not contained within the region.
+            If the co-ordinate is not contained within this part of the tree or
+            the core number is out of range.
 
         Returns
         -------
         bool
-            If all contained subregions are full.
+            True if the specified core is to be loaded to all subregions.
         """
         # Check that the co-ordinate is contained in this region
-        if ((x < self.base_x or x >= self.base_x + self.scale) or
+        if ((p < 0 or p > 17) or
+                (x < self.base_x or x >= self.base_x + self.scale) or
                 (y < self.base_y or y >= self.base_y + self.scale)):
-            raise ValueError((x, y))
+            raise ValueError((x, y, p))
 
         # Determine which subregion this refers to
         subregion = ((x >> self.shift) & 0x3) + 4*((y >> self.shift) & 0x3)
 
         if self.level == 3:
             # If level-3 then we just add to the locally selected regions
-            self.locally_selected.add(subregion)
+            self.locally_selected[p] |= 1 << subregion
         else:
```

Comment:
Should also check if the subregion has already been selected, otherwise the same core could be specified twice when calling getting regions and core masks out.